File: 0028_Implement_strStr()/Solution2.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if not needle:
            return 0
        if not haystack:
            return -1

        _len1, _len2 = len(haystack), len(needle)
        if _len2 > _len1:
            return -1

        p1, p2 = 0, 0
        while p1 < _len1 and p2 < _len2:
            if haystack[p1] == needle[p2]:
                p1 += 1
                p2 += 1
            else:
                p1 = p1 - p2 + 1    # 从 start 开始 p1 和 p2 走了一样的长度，减去这个长度再+1，即回到 start + 1
                p2 = 0

        if p2 != _len2:
            return -1
        return p1 - _len2


def test(test_name, haystack, needle, expected):
    res = Solution().strStr(haystack, needle)
    if res == expected:
        print(test_name + ' success.')
    else:
        print(test_name + ' failed.')

if __name__ == "__main__":
    haystack1 = "hello";
    needle1 = "ll";
    expected1 = 2;
    test("test1", haystack1, needle1, expected1);

    haystack2 = "aaaaa";
    needle2 = "bba";
    expected2 = -1;
    test("test2", haystack2, needle2, expected2);

    haystack3 = "";
    needle3 = "";
    expected3 = 0;
    test("test3", haystack3, needle3, expected3);

    haystack4 = "aaa"
    needle4 = "aaaa"
    expected4 = -1
    test('test4', haystack4, needle4, expected4)

    haystack5 = "a"
    needle5 = ""
    expected5 = 0
    test("test5", haystack5, needle5, expected5)

File: 0028_Implement_strStr()/Solution2.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int strStr(string haystack, string needle) {
        if (needle == "") return 0;
        if (haystack == needle) return 0;
        if (haystack.size() < needle.size()) return -1;

        int p1 = 0, p2 = 0;
        while (p1 < haystack.size() && p2 < needle.size()) {
            if (haystack[p1] == needle[p2]) {
                p1++; p2++;
            } else {
                p1 = p1 - p2 + 1;
                p2 = 0;
            }
        }
        if (p2 == needle.size())
            return p1 - needle.size();
        return -1;
    }
};

void test(string test_name, string haystack, string needle, int expected)
{
    Solution s;
    if (s.strStr(haystack, needle) == expected) {
        cout << test_name << " success." << endl;
    } else {
        cout << test_name << " failed." << endl;
    }
}

int main()
{
    string haystack1 = "hello";
    string needle1 = "ll";
    int expected1 = 2;
    test("test1", haystack1, needle1, expected1);

    string haystack2 = "aaaaa";
    string needle2 = "bba";
    int expected2 = -1;
    test("test2", haystack2, needle2, expected2);

    string haystack3 = "";
    string needle3 = "";
    int expected3 = 0;
    test("test3", haystack3, needle3, expected3);

    string haystack4 = "";
    string needle4 = "a";
    int expected4 = -1;
    test("test4", haystack4, needle4, expected4);

    string haystack5 = "a";
    string needle5 = "a";
    int expected5 = 0;
    test("test5", haystack5, needle5, expected5);

    string haystack6 = "aaa";
    string needle6 = "aaaa";
    int expected6 = -1;
    test("test6", haystack6, needle6, expected6);

    return 0;
}
